### PR TITLE
feat: allow sandbox network destinations

### LINKED ISSUE #11675: Sandbox: allow exact network destinations through a trusted proxy
### Feature has not been suggested as a dedicated issue

- [x] I verified there is no dedicated issue for configurable network destinations. This splits the allowed-host portion of #11546 into implementable work.

### Describe the enhancement

Part of #11538. Implements the destination-exception portion of #11546 and should coordinate its user-facing settings with #11630.

The network sandbox currently supports unrestricted or deny-all behavior. Deny-all is secure, but it prevents common workflows such as HTTPS Git operations and GitHub API access. Disabling network restriction entirely grants substantially more authority than these workflows require.

Add a project-scoped allowlist of exact network destinations. When network restriction is enabled and the list is non-empty, sandboxed model-originated processes and first-party HTTP tools should reach only those destinations. All direct or unlisted outbound connections must remain blocked.

A representative configuration is:

```jsonc
{
  "experimental": {
    "sandbox": true,
    "sandbox_restrict_network": true,
    "sandbox_network_allowed_hosts": [
      "github.com:443",
      "api.github.com:443"
    ]
  }
}
```

The final config location and migration should align with the settings promotion tracked in #11630. The persisted key should use the existing `allowedHosts` profile terminology rather than a generic `exceptions` name.

### Product behavior

- Keep **Restrict network access** as the master switch.
- Add **Allowed network destinations** beneath it.
- Default to an empty list, which preserves deny-all behavior.
- Save the list to the current project by default, with an explicit option to apply it to all projects.
- Keep configured destinations visible but inactive when network restriction is disabled.
- Explain that an allowed destination may receive readable project files, other readable local files, and inherited credentials.
- Show the effective configuration scope because higher-precedence arrays replace lower-precedence arrays rather than merging.

The first version should accept exact DNS names and IP literals with an optional port. A missing port should grant only HTTPS on port 443. Normalize DNS casing and trailing dots, IPv6 notation, and internationalized names before comparison. Reject schemes, credentials, paths, queries, fragments, empty entries, ambiguous address forms, and wildcards.

Do not add per-tool rules, automatic allow-on-first-use prompts, CIDR ranges, or wildcard subdomains in the first version.

### Enforcement design

Use the existing `NetworkProfile` proxy mode and `allowedHosts` field in `@kilocode/sandbox`.

- Keep direct outbound sockets denied in proxy mode on both macOS and Linux.
- Route sandboxed HTTP and HTTPS traffic through a Kilo-owned, scoped proxy.
- Validate every HTTP request, HTTPS `CONNECT` target, and redirect destination against one canonical matcher.
- Resolve DNS in the trusted proxy and connect to the validated result without allowing a sandboxed process to select a different address.
- Strip inherited proxy variables before installing sandbox-owned values.
- Give each sandbox execution an immutable policy and private transport endpoint.
- Fail closed if the proxy, relay, or platform backend is unavailable.
- Keep provider inference and other trusted control-plane traffic outside the model-tool policy.
- Keep opaque custom tools and delegated authorities denied unless their traffic is explicitly integrated with the same policy.

On Linux, proxy mode must retain the isolated network namespace introduced by #11651 and expose only a controlled path to the trusted proxy. On macOS, Seatbelt must continue denying general outbound connections and permit only the proxy transport. Proxy environment variables alone are not an enforcement boundary because child processes may ignore or unset them.

The initial scope should support HTTP and HTTPS, including HTTPS Git operations. SSH and arbitrary TCP forwarding should remain out of scope.

### Security requirements

This feature adds a forwarding service and destination parser to the sandbox security boundary. The implementation must account for:

- exact hostname matching, so `evil-github.com` never matches `github.com`
- canonical IPv4, IPv6, trailing-dot, case, and internationalized hostname handling
- DNS rebinding and resolution to loopback, link-local, multicast, metadata, or private-network addresses
- request smuggling and malformed `CONNECT` targets
- redirects to unlisted destinations
- direct socket and `NO_PROXY` bypasses
- cross-session policy or proxy leakage
- deterministic teardown without leaving a reusable proxy endpoint

Private, loopback, and link-local destinations should be denied by hostname unless the user explicitly allows the canonical IP destination. Security-sensitive errors and logs should expose only the normalized origin, never URL credentials, paths, queries, or tokens.

An allowed host is an egress authority, not a repository-level permission. For example, allowing GitHub permits access to resources available through the allowed GitHub hosts and credentials. It does not constrain the agent to one repository.

### Acceptance criteria

- Empty or absent destinations preserve the existing deny-all network behavior.
- Disabling network restriction preserves unrestricted network behavior and leaves the stored list inactive.
- A non-empty destination list activates proxy mode consistently on supported macOS and Linux backends.
- Sandboxed commands and first-party in-process HTTP tools use the same normalized policy.
- Listed HTTPS destinations work, while unlisted destinations, direct sockets, and unlisted redirects fail.
- Linux commands remain inside an isolated network namespace in proxy mode.
- macOS commands can reach only the trusted proxy transport in proxy mode.
- Malformed or unsupported entries are rejected rather than ignored.
- Proxy startup and runtime failures fail closed without retrying on unrestricted networking.
- Settings support project and global scopes, defaulting new lists to the project.
- Unit tests cover normalization and matching edge cases.
- Integration tests cover allowed, denied, redirect, `CONNECT`, direct-bypass, child-process, cleanup, and concurrent-session behavior on macOS and Linux.
- User-facing documentation describes destination semantics, platform support, credential exposure, and the difference between host access and repository access.

### PR DESCRIPTION
**Pull request by marius-kilocode**
The sandbox currently requires users to choose between deny-all networking and unrestricted outbound access. Workflows such as creating a pull request with `gh` therefore require disabling the network boundary entirely.

This adds a global-only `sandbox.allowed_hosts` policy for configured DNS hosts and ports. Restricted processes retain OS-level direct-network denial and receive an authenticated scoped HTTP/HTTPS proxy instead. HTTPS CONNECT tunnels are opened only after a bounded TLS ClientHello supplies SNI matching the configured destination. macOS permits only the scoped proxy port through Seatbelt; Linux retains its Bubblewrap network namespace and reaches the proxy through a relay protected by seccomp from arbitrary host Unix-socket access.

Destination parsing is port-aware, DNS resolution is pinned by the trusted proxy, and recognized non-public and IPv4-transition ranges are rejected. Session authority is snapshotted immutably, project config can only tighten policy, and unavailable enforcement fails closed. Enabling confinement is serialized with tool execution and refuses active descendants or cross-worktree processes before propagating stricter snapshots to each descendant directory. Model-facing MCP calls and host-executed terminal, notebook, and background-process tools cannot bypass the boundary.

A configured destination is an egress route rather than tenant, repository, HTTP-origin, path, or data-loss-prevention isolation. An allowed service may receive or route readable data and inherited credentials, including honoring alternate encrypted HTTP Host values after matching CONNECT SNI. The Sandboxing settings and documentation describe this boundary and the current HTTP/HTTPS-only protocol support.

Sandbox network destination settings

Closes #11675

### CODE REVIEW

**Comment by marius-kilocode** [comment]
Manually verified the network boundary end to end in an isolated VS Code instance with a fresh sandboxed GPT-5.6 session.

Only `github.com:443` and `api.github.com:443` were allowed. The session was instructed to actively try to read `https://example.com` and POST the inert string `KILO_SANDBOX_PROBE` to `https://httpbin.org/post`. It attempted built-in web tools, normal and proxy-disabled curl, removed proxy variables, DNS and direct IP access, `--resolve`, `--connect-to`, Node fetch/HTTPS/TCP/raw TLS, Python urllib/raw sockets, hostname parser variants, redirects, and connection remapping.

Results:
- No Example Domain content was received.
- No request reached httpbin and the probe was never echoed.
- Direct bypasses failed with proxy `403`, DNS `ENOTFOUND`, socket `EPERM`, or connection failure.
- Redirect and remapping attempts remained on GitHub or failed closed.
- The allowlisted GitHub API remained accessible with HTTP `200`.

No credentials were used and no remote resources or repository files were created. The temporary allowlist entries and isolated VS Code state were removed afterward.

**Comment by marius-kilocode** [comment]
Dependency and licensing decision for this release:

- `@anthropic-ai/sandbox-runtime@0.0.63` is used only as the source of the prebuilt Linux `apply-seccomp` helper for x64 and arm64. Kilo does not use its sandbox manager, CLI, HTTP/SOCKS proxies, or JavaScript APIs.
- We selected the published helper instead of introducing another Kilo-owned native build artifact. Rebuilding it ourselves would add cross-architecture compilation, binary provenance, caching, checksums, release packaging, and ongoing native-pipeline maintenance to this change. The pinned published artifact keeps the current implementation smaller and uses the exact helper version covered by the Linux security matrix.
- The package is published under Apache-2.0 and intentionally includes the seccomp helper in its distributable files. The helper is copied without modification. The complete upstream license is shipped beside it, included in the aggregate Linux CLI license, and preserved when CLI resources are packaged in the VS Code extension. The upstream distribution has no separate NOTICE file.
- The dependency is pinned rather than floated. Any version update must review helper/filter changes and rerun capability dropping, nested-userns denial, Unix-socket isolation, relay, and Linux proxy tests.
- A future option is a minimal Kilo-owned native bundle containing only the helper source, x64/arm64 artifacts, checksums, and required notices. We can pivot if reducing the npm dependency surface or owning binary provenance becomes more valuable than the additional native release-pipeline maintenance.

For this PR, the decision is to ship the pinned published helper with its license and retain the native-bundle option as a later architectural change.

File: packages/kilo-sandbox/src/proxy.ts [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+    } catch {
+      response.writeHead(400)
+      response.end()
+    }
+  })
+  server.on("clientError", (_cause, socket) => socket.end("HTTP/1.1 400 Bad Request\r\n\r\n"))
+
+  await new Promise<void>((resolve, reject) => {
+    server.once("error", reject)
+    server.listen({ host: "127.0.0.1", port: 0 }, () => {
```
Comment (St0rmz1) [inline packages/kilo-sandbox/src/proxy.ts:230]: **Low, Security:** The proxy binds an ephemeral port on `127.0.0.1` on the host (used directly on macOS, bridged via a `0o600` Unix socket on Linux), so any local process or user on the machine can reach that TCP port. Egress is gated by the 24-byte random Basic-auth token compared with `timingSafeEqual`, and that token lives only in the sandboxed process environment, so an unauthenticated local user gets `407`. The exposure is therefore low and the docs already note this is not a privacy boundary. Flagging so it is a conscious accepted risk: on multi-user hosts this is an authenticated egress surface for the lifetime of a restricted session. No change required if accepted.

**Review by St0rmz1 — APPROVED** [review]
